[RUNTIME] Adds UDP communication capabilities with Dawn

bodiddles opened the pull request:
Refactors UDP processes into two processes (send and receive). Each process
then spawns 2 threads, one for packaging/unpackaging, and one for sending/receiving
respectively.

The send and receive processes are now encapsulated in classes that 
the shared buffers between the package and send and the receive and unpackage
threads respectively. Each process class inherits from a super class that handles
thread creation and starting said threads.

Currently the code just sends a test string and adds the bytes received from
the receive port into the state dictionary with the key of "bytes". Protobuf
integration and still needs to be accomplished as well as handling gamepad
data from Dawn.

Adds 2 command types and corresponding functions for state manager to send
the state to the send process as well as adding the data received from Dawn
into the dictionary.

Creates new process name enums and adds thread enums for state manager
to keep track of all programs running in Ansible.

The current Ansible has been tested with fake_dawn.py.

bodiddles committed:
[RUNTIME] Adds basic UDP integration

Creates 4 processes: packager, sender, receiver, unpackager to handle
data and send over it over UDP sockets. These currently don't use
google protos but converts "TEST" into a bytearray and sends over sockets.

To handle async messages, there is a class called two_buffer which just
stores and sends data in two separate indices of a list which are atomic
operations which should be ok for async.

It uses the process factory in runtime.py to generate said processes and
I have added them into the process names enum class. I have also added
additional SM_COMMANDS SEND and RECEIVE. SEND is sent to SM when the
packager is done with packaging one set of data and can accept another.
This should be fixed in the future to constantly send over the state,
potentially using another instance of the two_buffer.
The RECEIVE command just stores whatever we send to state manager into
the state itself which will also need to fixed to handle storing gamepad
data which we expect from Dawn over UDP.

brandonxxlee committed:
[RUNTIME] Fix stylistic errors on UDP integration

bodiddles committed:
Fixed Style Changes recommended by reviewers

bodiddles committed:
[RUNTIME] Refactored Ansible to use classes

Adds AnsibleHandler class which can create threads with correct inputs
for both udp sender and udp receiver processes. There are also udpSendClass
and udpRecvClass which inherit these methods from AnsibleHandler and
adds the correct inputs to create said classes.

Also refactored runtime.py to account for this change and added two
methods to mirror startStateManager but for udp sender and receiver

Added thread names and changed process names in runtimeUtil to reflect
these changes

bodiddles committed:
[RUNTIME] Reformat Ansible.py into seperate classes

The functions used by the classes to send/receive UDP data is now placed within
the respective classes. Code needs documentation and testing.

bodiddles committed:
[RUNTIME] Reformat Ansible.py into seperate classes

The functions used by the classes to send/receive UDP data is now placed within the respective classes. Code needs documentation and testing.

brandonxxlee committed:
[RUNTIME] Create docstrings in Ansible.py

Create docstrings for half the functions and fix name errors.

bodiddles committed:
[RUNTIME] Completed refactor of ansible to classes

Desperately need to fix bugs

brandonxxlee committed:
[RUNTIME] Commit changes to merge with master

brandonxxlee committed:
[RUNTIME] Merge remote-tracking branch 'origin/master' into UDPnoproto

Conflicts:
	runtime/testy/runtimeUtil.py
	runtime/testy/stateManager.py
	runtime/testy/studentCode.py

brandonxxlee committed:
[RUNTIME] Add setup and main in studentCode

brandonxxlee committed:
[RUNTIME] Finish refactoring of Ansible

-Runtime and Ansible now works without protos with fake_dawn
-Next Step: integerate with real Dawn

brandonxxlee committed:
[RUNTIME] Fix minor bugs in statemanager

- SendUDP in StateManager now properly takes in zero arguments
- Change loop from pass to time.sleep in Ansible to increase performance, this
  is simply to keep the process running

bodiddles committed:
[Runtime] Finished UDP refactoring

Refactors UDP processes into two processes (send and receive). Each process
then spawns 2 threads, one for packaging/unpackaging, and one for sending/receiving
respectively.

The send and receive processes are now encapsulated in classes that creates
the shared buffers between the package and send and the receive and unpackage
threads respectively. Each process class inherits from a super class that handles
thread creation and starting said threads.

Currently the code just sends a test string and adds the bytes received from
the receive port into the state dictionary with the key of "bytes". Protobuf
integration and still needs to be accomplished as well as handling gamepad
data from Dawn.

Adds 2 command types and corresponding functions for state manager to send
the state to the send process as well as adding the data received from Dawn
into the dictionary.

Creates new process name enums and adds thread enums for state manager
to keep track of all programs running in Ansible.

The current Ansible has been tested with fake_dawn.py.

bodiddles committed:
[RUNTIME] Merged with master for PR

brandonxxlee committed:
[RUNTIME] Delete old files used before refactoring

yizow commented on runtime/testy/runtimeUtil.py:
```diff
@@ -19,6 +21,15 @@ class PROCESS_NAMES(Enum):
+  UDP_SENDER          = "udpSender"
+  UDP_RECEIVER        = "udpReceiver"
+  UDP_UNPACKAGER      = "udpUnpackager"
   HIBIKE              = "hibike"
```
Hibike isn't a thread.
my guess is somewhere in the merging you mixed this up.

yizow commented on runtime/testy/runtimeUtil.py:
```diff
@@ -19,6 +21,15 @@ class PROCESS_NAMES(Enum):
+  UDP_RECEIVE_PROCESS = "udpReceiveProcess"
+
+@unique
+class THREAD_NAMES(Enum):
```
Is there a reason to put this in runtimeUtil.py and not Ansible.py? 
I've been using runtimeUtil for things that are used across files. It looks like the thread names are only used within Ansible.py though. 

If so, i think putting it in there will make things cleaner and more maintaneable. Better abstraction, etc.

yizow commented on runtime/testy/runtime.py:
```diff
@@ -208,4 +219,4 @@ def startHibike(badThingsQueue, stateQueue, pipe):
 
 if __name__ == "__main__":
-  runtimeTest()
+  runtime()
```
I think the suite of tests that runtimeTest runs should be default behaviour.

yizow replied:
I still think runtimeTest should be run here.

yizow commented on runtime/testy/runtime.py:
```diff
@@ -118,6 +115,20 @@ def startStateManager(badThingsQueue, stateQueue, runtimePipe):
+    sendClass = Ansible.UDPSendClass(badThingsQueue, stateQueue, smPipe)
+    sendClass.start()
+  except Exception:
+    badThingsQueue.put(BadThing(sys.exc_info(), None))
```
Please put additional information in this bad thing. 

`BadThing(sys.exc_info(), "String description", event=BAD_EVENTS.EVENT_TYPE)

yizow commented on runtime/testy/runtime.py:
```diff
@@ -118,6 +115,20 @@ def startStateManager(badThingsQueue, stateQueue, runtimePipe):
+    recvClass = Ansible.UDPRecvClass(badThingsQueue, stateQueue, smPipe)
+    recvClass.start()
+  except Exception:
+    badThingsQueue.put(BadThing(sys.exc_info(), None))
```
See above

yizow commented on runtime/testy/fake_dawn.py:
```diff
@@ -0,0 +1,35 @@
+import threading
+import queue
+import time
+data = [0]
```
add newline between imports and global variables.

yizow commented on runtime/testy/fake_dawn.py:
```diff
@@ -0,0 +1,35 @@
+			b.extend(map(ord, string))
+			msg = bytes(b)
+			s.sendto(msg, (host, send_port))
+def receiver(port, receive_queue):
```
add newline between functions

yizow commented on runtime/testy/fake_dawn.py:
```diff
@@ -0,0 +1,35 @@
+recv_thread.daemon = True
+recv_thread.start()
+sender_thread.start()
+while True:
```
Newline before thread initialization and 'main' function

yizow replied:
This is busy waiting. This is also generally really bad for performance. 

I suppose if you are using this to simulate Dawn, it should be fine. Since it will be replaced, right?

yizow commented on runtime/testy/fake_dawn.py:
```diff
@@ -0,0 +1,35 @@
+sender_thread.start()
+while True:
+	print(data)
+	time.sleep(1)
```
Add newline at end of file.

yizow commented on runtime/testy/fake_dawn.py:
```diff
@@ -0,0 +1,35 @@
+recv_port = 1235
+
+def sender(port, send_queue):
+	host = socket.gethostname()
```
Can probably move `host` assignment to outside both of these functions. The host should never change while we are running the threads.

yizow commented on runtime/testy/fake_dawn.py:
```diff
@@ -0,0 +1,35 @@
+data = [0]
+send_port = 1236
+recv_port = 1235
+
```
Why create two different sockets? Shouldn't we be using the same socket to both send and receive information, since it's the same connection? 

After some internet digging, it seems using SOCK_DGRAMS means that the send and recv on the socket are atomic. I think it should be fine.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+        socketThread = self.threadMaker(self.socketFunc, self.socketName)
+        packagerThread.start()
+        socketThread.start()
+        while True:
```
This is busy waiting. 
This is bad.

Use `packagerThread.join(); socketThread.join()`

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+        by the package function, defined internally. The packaged data is then placed 
+        back into the TwoBuffer replacing the previous state.
+        """
+        def package(state, badThingsQueue):
```
You can remove the badThingsQueue object. It is already bound in the scope of the parent object. (Assuming there is only one badThingsQueue object being shared within the scope of packageData)

yizow replied:
Do we need to have the badThingsQueue as an argument?

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+    Initializes generalized instance variables for both UDP sender and receiver, and creates a callable
+    method to initialize the two threads per UDP process and start them. 
+    """
+    def __init__(self, packagerName, socketName, packager, socketThread, badThingsQueue, stateQueue, pipe):
```
might make more sense to organize the argument list to be `packagerName, packagerThread, socketName, socketThread` to improve clarity.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+            eventually be implemented to package the rawState into protos.
+            """
+            s = "TEST" 
+            b = bytearray()
```
The extend function can be combined with initialization. Both take an iterable of integers in the range [0, 256) as arguments.
`b = bytearray(map(ord, s))`

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+            try:
+                stateQueue.put([SM_COMMANDS.SEND_ANSIBLE, []])
+                rawState = pipe.recv()
+                if rawState == RUNTIME_CONFIG.PIPE_READY:
```
These two lines will never be received by you. Deprecated behaviour, please remove.

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+        while True:
+            try:
+                unpackagedData = unpackage(self.recvBuffer.get())
+                stateQueue.put([SM_COMMANDS.RECV_ANSIBLE, [unpackagedData]])
```
This loop will keep unpacking the same data repeatedly, wasting CPU time. Don't do this.

nikitakit reviewed (changes requested):
Sender class looks fine, but the receiver feels wrong.

You shouldn't need multiple thread for the receiver. Just do socket read, unpackage, followed by adding to statemanager all in the same thread. Also don't use a TwoBuffer for receive.

If you're worried about not being able to keep up with received data, you can do that by asking the kernel how much data is in the socket, and dropping some packets if the kernel reports a backlog. But this is an optimization that can be done in a later PR (it might be unneeded anyway).

The use of multiple threads (a) adds latency at each cross-thread hop, (b) currently uses a busy-loop that will steal all of the BBB's CPU time, and (c) is actually more complex than the single-threaded approach.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+                rawState = pipe.recv()
+                if rawState == RUNTIME_CONFIG.PIPE_READY:
+                    stateQueue.put([SM_COMMANDS.SEND_ANSIBLE, []])
+                elif rawState:
```
Assuming rawState is a dictionary:
An empty dictionary returns a false value. Are we sure than an empty dictionary will _never_ have semantic meaning? 

Does this if check guard against anything? Even if we receive an empty dictionary, could there be a case where we actually do want to send an empty dictionary? 
Consider just removing the if statement altogether. we're calling `pipe.recv()`, meaning that stateManager has deliberately put something on the pipe for us to see.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+                stateQueue.put([SM_COMMANDS.SEND_ANSIBLE, []])
+                rawState = pipe.recv()
+                if rawState == RUNTIME_CONFIG.PIPE_READY:
+                    stateQueue.put([SM_COMMANDS.SEND_ANSIBLE, []])
```
Do we perhaps want to limit how fast this runs? Currently, we are asking for new info to send as soon as we send the previous one. Could we flood the queue so that nothing else gets processed? 

If you want to rate limit this, look in runtime.py for how i'm rate limiting the frequency of studentCode.main() calls.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+        """
+        host = socket.gethostname() #TODO: determine host in runtime-dawn comm
+        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
+            while True: 
```
This loop needs a time.sleep in it somewhere. Currently, this is just constantly sending without any rate limiting. 

time.sleep at some set interval configurable in runtime.util would be my recommendation.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,176 @@
+                    "UDP sender thread has crashed with error:",  
+                    event = BAD_EVENTS.UDP_RECV_ERROR, 
+                    printStackTrace = True))
+                    
```
missing newline

yizow reviewed (changes requested):
Sorry for the delay. Here's a host of comments.

I agree with nikita's comment's on receiver, so didn't code review those parts.

bodiddles committed:
[RUNTIME] Fixed most issues brought up.

Fixed certain stylistic issues, added limiting to all sub-threads in Ansible.

Need to fix busy waiting in receiving and dropping packets with selects.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,188 @@
+            try:
+                stateQueue.put([SM_COMMANDS.SEND_ANSIBLE, []])
+                rawState = pipe.recv()
+                packState = package(rawState, badThingsQueue)
```
Dont need to pass in badThingsQueue. It's already in the scope of the package function.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,188 @@
+                rawState = pipe.recv()
+                packState = package(rawState, badThingsQueue)
+                self.sendBuffer.replace(packState)
+                time.sleep(1.0/self.packagerHZ)
```
Not accurate. Time varies depending on how long we block and how long the package function takes to execute. And on when the scheduler decides to run this process.
runtime.py has an example to tie the sleep duration to the amount of time the it takes to run.

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,188 @@
+                    msg = self.sendBuffer.get()
+                    if msg != 0: 
+                        s.sendto(msg, (host, UDPSendClass.SEND_PORT))
+                    time.sleep(1.0/self.socketHZ)
```
See comments above. Not necessarily accurate timing.

bodiddles committed:
changed limiting of ansible threads to better reflect actual timings

yizow commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,200 @@
+        self.pipe = pipe
+        self.packagerName = packagerName
+        self.socketName = socketName
+        self.packagerHZ = packagerHZ
```
Should these two be the same, or different? Will we ever want a case where one of these threads is running faster than the other?

yizow commented on runtime/testy/fake_dawn.py:
```diff
@@ -0,0 +1,39 @@
+def sender(port, send_queue):
+	host = '127.0.0.1'
+	with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
+		while True:
```
Isn't this going to flood the receiving process?

bodiddles replied:
Done

bodiddles committed:
adds limiting to the send thread in fake_dawn

brandonxxlee committed:
[RUNTIME] runtime.py now runs main and test

brandonxxlee committed:
[RUNTIME] Change HZ constant to be class variable

yizow committed:
[RUNTIME] UDP receive in single thread

Refactors receiving from Dawn over UDP to use python's selector class.
There is now a single thread that listens for new data from Dawn. When
new data is received, we read until there is no more data, then
unpackage the last read data, unpackage it, and send it to stateManager.

yizow committed:
Merge branch 'runtime/UDPnoproto' of github.com:pioneers/PieCentral into runtime/UDPnoproto

Conflicts:
	runtime/testy/Ansible.py

bodiddles committed:
fixed some syntax issues with select functionality

bodiddles committed:
finished refactoring with new schema in receive process

bodiddles committed:
added documentation to the new receive process start method

yizow reviewed (approved):
Things look good to me.

yizow commented:
Oh, make sure to squash and write a new commit message.
May want to mention brandon's name in the commit message as well, so people know there are two people who can answer questions on this code.

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,199 @@
+
+    """
+    def __init__(self):
+        self.data = [0, 0]
```
I believe Python convention is to use None instead of 0

nikitakit commented on runtime/testy/Ansible.py:
```diff
@@ -0,0 +1,199 @@
+                rawState = pipe.recv()
+                packState = package(rawState)
+                self.sendBuffer.replace(packState)
+                nextCall += 1.0/self.packagerHZ
```
This type of logic is not resilient against a one-time delay. If, say, package takes forever a single time, it is possible that this loop will never sleep again in the immediate aftermath.

bodiddles replied:
Doess the initialization of nextCall each loop not catch this as it only times how long each iteration of the loop takes? If it does take too long, it would never sleep but that would be the correct behavior as we only want to limit the maximum rate it goes right?

nikitakit replied:
You're right, I misread the code.

nikitakit reviewed (approved):
Major concerns have been addressed, and this now looks ready to merge!

I haven't actually run the code, so I trust that you guys actually tested that it works.

At this point integration takes priority over perfection. However, there are a few minor nitpicks I pointed out that you might want to consider at some point in the future.
